### PR TITLE
Issue multiple dummy nodes formed

### PR DESCRIPTION
**Pull request by himanshudube97**
Issues-> 
1. Click on source node (table ) then click on operational node, the column data in the operational form would dissapper.
reason-> It was a dummy operational node (and action was edit), so api call wont fetch any data with that nodeid.
sol-> If its an operational dummy node, keep it in `create` mode (not edit) and use source node id for api call.

2. Click on source node then any operation from operations table, a dummy node appears. Now again click source node (table) and choose another operation , a new dummy node appears and so on. 
reason-> No code to check if the dummy node created earlier is being used or not. 
sol-> If user clicks on source node , then remove the dummy node. This ensures we dont have multiple dummy node on the canvas.

Before: ->
[link1](https://jam.dev/c/c23da7a4-a82e-423c-bcbf-86ed084cd19d)

After:->
[link2](https://jam.dev/c/8334c6d4-ff0a-4ca7-baac-76a36bad2e7c)

### CODE REVIEW

File: src/components/TransformWorkflow/FlowEditor/Components/Canvas.tsx [hunk 1]
```diff
@@ -288,10 +289,10 @@ const Canvas = ({
   setTempLockCanvas,
 }: CanvasProps) => {
   const { data: session } = useSession();
-  const [nodes, setNodes, onNodesChange] = useNodesState([]);
-  const [edges, setEdges, onEdgesChange] = useEdgesState([]);
+  const [nodes, setNodes, onNodesChange] = useNodesState([]); //works when we click the node or move it.
+  const [edges, setEdges, onEdgesChange] = useEdgesState([]); //workds when we click the edges.
   const [openOperationConfig, setOpenOperationConfig] =
-    useState<boolean>(false);
+    useState<boolean>(false); // this is the left form with sql operations.
```
Comment (Ishankoradia) [inline src/components/TransformWorkflow/FlowEditor/Components/Canvas.tsx:295]: you mean the right hand side config panel right ?

File: src/components/TransformWorkflow/FlowEditor/Components/OperationPanel/Forms/CastColumnOpForm.tsx [hunk 2]
```diff
@@ -197,7 +204,16 @@ const CastColumnOp = ({
...
+  /**
+    So operation nodes can be dummy (not yet saved to db) or real (saved in db).
+    Both have nodeId, but the dummy nodes have isDummy= true field. 
+    So we dont call any api in that case.
+   **/
   useEffect(() => {
+    if (node?.type === SRC_MODEL_NODE) {
+      sourceModelNodeRef.current = node; //saving table's id
+    }
+    if (node?.data.isDummy) return;
```
Comment (Ishankoradia) [inline src/components/TransformWorkflow/FlowEditor/Components/OperationPanel/Forms/CastColumnOpForm.tsx:216]: this works for cast operation but not others. Also this doesn't handle the case when the node is in the middle of the chain

Try this for a table -> create a drop operation and save -> add function -> select the filter operation , you will the see columns in drop down to filter on -> click on the dummy node of filter -> check the columns in filter operations's dropdown again.

**Comment by himanshudube97** [comment]
Issues->
1. When clicked on a dummy operational node the data in the right form would go empty.
2. If there was a dummy node created and then again we click on the parent node (source /operational) , another dummy node would comeup on screen. So this way multiple dummy nodes were there.

Changes made->
1. added a parent node as a field in the dummy node. So that we know from which parent that node is created. This works for both if the parent node is a sourceNode(table) or an operationalNode.

2. Also a `custom hook` `useOpForm` has been added. Right now it is returning the nodeData and the parentNode. In next iterations we will add the inital useeffect that fetches data (common code in all components).

3. In `operationalConfigLayout`, code to check if there is a dummy node and user clicks on some other node, than the dummy node is deleted.

4. In `Canvas.tsx` file its a minor ui change of the top `run workflow `button.

Comman Changes in all forms:->

1. When the forms are rendering, and if the node is dummy node, then useEffect don't call the apis.

2. While Saving: 
   a. checking if the node is a dummy node or a real node (saved in the db).
   b. if its a real node it have a id (saved in db), works fine, else if its a dummy node, we take id and other data from its parent, and mark this operation as create not edit (for dummy nodes).

**Comment by himanshudube97** [comment]
Issues still remaining->
1. Cases where the dummy node is a source node (table), eg `join operation`. When clicked on the dummy source node created , it resets the form.

**Review by Ishankoradia — APPROVED** [review]

**Pull request merged**